gctscript/vm/vm.go

```diff
@@ -0,0 +1,284 @@
+}
+
+// Compile compiles to byte code loaded copy of vm script
+func (vm *VM) Compile() (err error) {
+	vm.Compiled = new(tengo.Compiled)
+	vm.Compiled, err = vm.Script.Compile()
+	return
+}
+
+// Run runs byte code
+func (vm *VM) Run() (err error) {
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Running script: %s ID: %v", vm.ShortName(), vm.ID)
+	}
+
+	err = vm.Compiled.Run()
+	if err != nil {
+		vm.event(StatusFailure, TypeExecute, true)
+		return Error{
+			Action: "Run",
+			Cause:  err,
+		}
+	}
+	vm.event(StatusSuccess, TypeExecute, true)
+	return
+}
+
+// RunCtx runs compiled byte code with context.Context support.
+func (vm *VM) RunCtx() (err error) {
+	if vm.ctx == nil {
+		vm.ctx = context.Background()
+	}
+
+	ct, cancel := context.WithTimeout(vm.ctx, GCTScriptConfig.ScriptTimeout)
+	defer cancel()
+
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Running script: %s ID: %v", vm.ShortName(), vm.ID)
+	}
+
+	err = vm.Compiled.RunContext(ct)
+	if err != nil {
+		vm.event(StatusFailure, TypeExecute, true)
+		return Error{
+			Action: "RunCtx",
+			Cause:  err,
+		}
+	}
+	vm.event(StatusSuccess, TypeExecute, true)
+	return
+}
+
+// CompileAndRun Compile and Run script with support for task running
+func (vm *VM) CompileAndRun() {
+	err := vm.Compile()
+	if err != nil {
+		log.Error(log.GCTScriptMgr, err)
+		err = RemoveVM(vm.ID)
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+		}
+		return
+	}
+
+	err = vm.RunCtx()
+	if err != nil {
+		log.Error(log.GCTScriptMgr, err)
+		err = RemoveVM(vm.ID)
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+		}
+		return
+	}
+	if vm.Compiled.Get("timer").String() != "" {
+		fmt.Println("Timer: ", vm.Compiled.Get("timer").String())
+		vm.T, err = time.ParseDuration(vm.Compiled.Get("timer").String())
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+			err = vm.Shutdown()
+			if err != nil {
+				log.Error(log.GCTScriptMgr, err)
+			}
+			return
+		}
+		if vm.T < time.Nanosecond {
+			log.Error(log.GCTScriptMgr, "Repeat timer cannot be under 1 nano second")
+			err = vm.Shutdown()
+			if err != nil {
+				log.Errorln(log.GCTScriptMgr, err)
+			}
+			return
+		}
+		vm.runner()
+	} else {
+		err = vm.Shutdown()
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+		}
+		return
+	}
+}
+
+// Shutdown shuts down current VMP
+func (vm *VM) Shutdown() error {
+	if vm.S != nil {
+		close(vm.S)
+	}
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Shutting down script: %s ID: %v", vm.ShortName(), vm.ID)
+	}
+	vm.Script = nil
+	pool.Put(vm.Script)
+	vm.event(StatusSuccess, TypeStop, true)
+	return RemoveVM(vm.ID)
+}
+
+func (vm *VM) Read() ([]byte, error) {
+	vm.event(StatusSuccess, TypeRead, true)
+	return vm.read()
+}
+
+// Read contents of script back
+func (vm *VM) read() ([]byte, error) {
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Read script: %s ID: %v", vm.ShortName(), vm.ID)
+	}
+	return ioutil.ReadFile(vm.File)
```

Comment:
As discussed, in that case of a new VM (when using execute), this will error out when execution type is TypeCreate